File: app/src/main/java/com/langualens/app/translate/Languages.kt
package com.langualens.app.translate

/** The 59 languages ML Kit can translate fully on device, plus the app's own interface locales. */
object Languages {

    data class Lang(val tag: String, val english: String, val native: String) {
        val label: String get() = if (english == native) english else "$native  ·  $english"
    }

    /** Shown at the top of the picker because they are the ones actually used here. */
    private val PINNED = listOf("nl", "en", "hu", "de", "es", "fr", "mk", "hr", "it", "pl")

    val ALL: List<Lang> = listOf(
        Lang("af", "Afrikaans", "Afrikaans"),
        Lang("ar", "Arabic", "العربية"),
        Lang("be", "Belarusian", "Беларуская"),
        Lang("bg", "Bulgarian", "Български"),
        Lang("bn", "Bengali", "বাংলা"),
        Lang("ca", "Catalan", "Català"),
        Lang("cs", "Czech", "Čeština"),
        Lang("cy", "Welsh", "Cymraeg"),
        Lang("da", "Danish", "Dansk"),
        Lang("de", "German", "Deutsch"),
        Lang("el", "Greek", "Ελληνικά"),
        Lang("en", "English", "English"),
        Lang("eo", "Esperanto", "Esperanto"),
        Lang("es", "Spanish", "Español"),
        Lang("et", "Estonian", "Eesti"),
        Lang("fa", "Persian", "فارسی"),
        Lang("fi", "Finnish", "Suomi"),
        Lang("fr", "French", "Français"),
        Lang("ga", "Irish", "Gaeilge"),
        Lang("gl", "Galician", "Galego"),
        Lang("gu", "Gujarati", "ગુજરાતી"),
        Lang("he", "Hebrew", "עברית"),
        Lang("hi", "Hindi", "हिन्दी"),
        Lang("hr", "Croatian", "Hrvatski"),
        Lang("ht", "Haitian Creole", "Kreyòl ayisyen"),
        Lang("hu", "Hungarian", "Magyar"),
        Lang("id", "Indonesian", "Bahasa Indonesia"),
        Lang("is", "Icelandic", "Íslenska"),
        Lang("it", "Italian", "Italiano"),
        Lang("ja", "Japanese", "日本語"),
        Lang("ka", "Georgian", "ქართული"),
        Lang("kn", "Kannada", "ಕನ್ನಡ"),
        Lang("ko", "Korean", "한국어"),
        Lang("lt", "Lithuanian", "Lietuvių"),
        Lang("lv", "Latvian", "Latviešu"),
        Lang("mk", "Macedonian", "Македонски"),
        Lang("mr", "Marathi", "मराठी"),
        Lang("ms", "Malay", "Bahasa Melayu"),
        Lang("mt", "Maltese", "Malti"),
        Lang("nl", "Dutch", "Nederlands"),
        Lang("no", "Norwegian", "Norsk"),
        Lang("pl", "Polish", "Polski"),
        Lang("pt", "Portuguese", "Português"),
        Lang("ro", "Romanian", "Română"),
        Lang("ru", "Russian", "Русский"),
        Lang("sk", "Slovak", "Slovenčina"),
        Lang("sl", "Slovenian", "Slovenščina"),
        Lang("sq", "Albanian", "Shqip"),
        Lang("sv", "Swedish", "Svenska"),
        Lang("sw", "Swahili", "Kiswahili"),
        Lang("ta", "Tamil", "தமிழ்"),
        Lang("te", "Telugu", "తెలుగు"),
        Lang("th", "Thai", "ไทย"),
        Lang("tl", "Tagalog", "Tagalog"),
        Lang("tr", "Turkish", "Türkçe"),
        Lang("uk", "Ukrainian", "Українська"),
        Lang("ur", "Urdu", "اردو"),
        Lang("vi", "Vietnamese", "Tiếng Việt"),
        Lang("zh", "Chinese", "中文")
    )

    private val byTag: Map<String, Lang> = ALL.associateBy { it.tag }

    /** Pinned entries first, then the rest alphabetically by English name. */
    val ORDERED: List<Lang> by lazy {
        val pinned = PINNED.mapNotNull { byTag[it] }
        val rest = ALL.filter { it.tag !in PINNED }.sortedBy { it.english }
        pinned + rest
    }

    fun find(tag: String): Lang? = byTag[tag]

    fun nameOf(tag: String): String = byTag[tag]?.native ?: tag.uppercase()

    /**
     * Locales the app interface itself is translated into.
     * An empty tag means "follow the system language".
     */
    val INTERFACE: List<Lang> = listOf(
        Lang("", "System default", "System default"),
        Lang("en", "English", "English"),
        Lang("nl", "Dutch", "Nederlands"),
        Lang("hu", "Hungarian", "Magyar"),
        Lang("de", "German", "Deutsch"),
        Lang("es", "Spanish", "Español"),
        Lang("fr", "French", "Français"),
        Lang("mk", "Macedonian", "Македонски"),
        Lang("sr-Latn", "Serbian", "Srpski")
    )

    fun interfaceName(tag: String): String =
        INTERFACE.firstOrNull { it.tag == tag }?.native ?: tag
}

File: extension/languages.js
/* Shared language list. Mirrors Languages.kt in the Android app so both halves
 * of LanguaLens offer the same pairs. Loaded as a plain script by the content
 * script and imported as a module by the popup and the service worker. */
(function (root) {
  var PINNED = ['nl', 'en', 'hu', 'de', 'es', 'fr', 'mk', 'hr', 'it', 'pl'];

  var ALL = [
    ['af', 'Afrikaans', 'Afrikaans'], ['ar', 'Arabic', 'العربية'],
    ['be', 'Belarusian', 'Беларуская'], ['bg', 'Bulgarian', 'Български'],
    ['bn', 'Bengali', 'বাংলা'], ['ca', 'Catalan', 'Català'],
    ['cs', 'Czech', 'Čeština'], ['cy', 'Welsh', 'Cymraeg'],
    ['da', 'Danish', 'Dansk'], ['de', 'German', 'Deutsch'],
    ['el', 'Greek', 'Ελληνικά'], ['en', 'English', 'English'],
    ['eo', 'Esperanto', 'Esperanto'], ['es', 'Spanish', 'Español'],
    ['et', 'Estonian', 'Eesti'], ['fa', 'Persian', 'فارسی'],
    ['fi', 'Finnish', 'Suomi'], ['fr', 'French', 'Français'],
    ['ga', 'Irish', 'Gaeilge'], ['gl', 'Galician', 'Galego'],
    ['gu', 'Gujarati', 'ગુજરાતી'], ['he', 'Hebrew', 'עברית'],
    ['hi', 'Hindi', 'हिन्दी'], ['hr', 'Croatian', 'Hrvatski'],
    ['ht', 'Haitian Creole', 'Kreyòl ayisyen'], ['hu', 'Hungarian', 'Magyar'],
    ['id', 'Indonesian', 'Bahasa Indonesia'], ['is', 'Icelandic', 'Íslenska'],
    ['it', 'Italian', 'Italiano'], ['ja', 'Japanese', '日本語'],
    ['ka', 'Georgian', 'ქართული'], ['kn', 'Kannada', 'ಕನ್ನಡ'],
    ['ko', 'Korean', '한국어'], ['lt', 'Lithuanian', 'Lietuvių'],
    ['lv', 'Latvian', 'Latviešu'], ['mk', 'Macedonian', 'Македонски'],
    ['mr', 'Marathi', 'मराठी'], ['ms', 'Malay', 'Bahasa Melayu'],
    ['mt', 'Maltese', 'Malti'], ['nl', 'Dutch', 'Nederlands'],
    ['no', 'Norwegian', 'Norsk'], ['pl', 'Polish', 'Polski'],
    ['pt', 'Portuguese', 'Português'], ['ro', 'Romanian', 'Română'],
    ['ru', 'Russian', 'Русский'], ['sk', 'Slovak', 'Slovenčina'],
    ['sl', 'Slovenian', 'Slovenščina'], ['sq', 'Albanian', 'Shqip'],
    ['sv', 'Swedish', 'Svenska'], ['sw', 'Swahili', 'Kiswahili'],
    ['ta', 'Tamil', 'தமிழ்'], ['te', 'Telugu', 'తెలుగు'],
    ['th', 'Thai', 'ไทย'], ['tl', 'Tagalog', 'Tagalog'],
    ['tr', 'Turkish', 'Türkçe'], ['uk', 'Ukrainian', 'Українська'],
    ['ur', 'Urdu', 'اردو'], ['vi', 'Vietnamese', 'Tiếng Việt'],
    ['zh', 'Chinese', '中文']
  ].map(function (row) {
    return { tag: row[0], english: row[1], native: row[2] };
  });

  var byTag = {};
  ALL.forEach(function (l) { byTag[l.tag] = l; });

  var ORDERED = PINNED.map(function (t) { return byTag[t]; })
    .filter(Boolean)
    .concat(
      ALL.filter(function (l) { return PINNED.indexOf(l.tag) < 0; })
        .sort(function (a, b) { return a.english < b.english ? -1 : 1; })
    );

  root.LLLanguages = {
    ALL: ALL,
    ORDERED: ORDERED,
    nameOf: function (tag) { return byTag[tag] ? byTag[tag].native : String(tag).toUpperCase(); },
    labelOf: function (tag) {
      var l = byTag[tag];
      if (!l) { return String(tag).toUpperCase(); }
      return l.english === l.native ? l.english : l.native + '  ·  ' + l.english;
    }
  };
})(typeof self !== 'undefined' ? self : this);
